[JS] Documentaiton cache and register destination (#667)
SAP/cloud-sdk: docs/js/features/connectivity/destination-cache-isolation.mdx

Examples of such a flow are `BasicAuthentication`, `ClientCertificateAuthentication` or `OAuth2ClientCredentials`.

There are cases where the choice by the SAP Cloud SDK is not optimal.
Assume you have a multi-tenant scenario and you need to pass the `JWT` to obtain the destination for the right tenant.
The `JWT` contains a `user_id`  but the destination flow is user independent.
The SAP Cloud SDK would use isolation `UserTenant`, although `Tenant` would be sufficient.
In such a case you can manually enforce a weaker isolation:

 <Tabs groupId="version" defaultValue="major" values={[
 { label: 'SDK 1.x', value: 'deprecated'},
 { label: 'SDK 2.x', value: 'major'}]}>

 <TabItem value="major">

 ```ts
 .execute({destinationName: 'myDestination', jwt: 'yourJWT', isolationStrategy: IsolationStrategy.Tenant})
 ```

 </TabItem>
 <TabItem value="deprecated">

 ```ts
 .execute({ destinationName: 'myDestination', jwt: 'yourJWT' }, { useCache: true, isolationStrategy: IsolationStrategy.Tenant })
 ```

 </TabItem>
 </Tabs>

Be very careful with this option.
If you manually reduce the isolation be sure that your destination flow is really not propagating a user to the target system.

> A erroneous configuration can lead to severe permission escalation between users.

Reviewer: I agree, we can leave exceptionally out. I tried to follow your style by bringing attention to this situation.
Should we also make it an admonition?

> :::Attention

Reviewer: ```suggestion
:::caution
```

> In version 1.X of the SAP Cloud SDK the isolation strategy is always TenantUser per default and you need to set it manually.

Reviewer: ```suggestion
In version 1.X of the SAP Cloud SDK the isolation strategy is always TenantUser by default. You need to set it manually.
```

Reviewer: I don't understand what you wanted to say here. Why set it manually if it's already set by default?

> Also you need to switch on the cache explicitly even if you pass an isolation strategy.

Reviewer: ```suggestion
You need to activate caching explicitly even if you pass an isolation strategy.
```

Reviewer: Maybe a link on how to activate it or a code snippet?

:::

## Register-Destination Cache

> Version 2.0 introduced the `registerDestination()` method, which is like a manual cache for destinations.
> We explain the usage of this method in the guide on [destination lookup](./destination.mdx).
> The method behaves similar to the destination-service cache discussed above.

Reviewer: ```suggestion
Version 2.0 introduced the `registerDestination()` method, which is like a manual cache for destinations.
We explain the usage of this method in the guide on [destination lookup](./destination.mdx).
The method behaves similarly to the destination-service cache discussed above.
```

```ts
import { registerDestination } from '@sap-cloud-sdk/connectivity';

const destination = {
    name: 'FORWARD-DESTINATION',
    url: 'https://mys4hana.com'
}

registerDestination(destination,{jwt:'yourJWT'});
```

> The chosen isolation depends on the given `JWT`:

Reviewer: ```suggestion
The isolation strategy is automatically detected and applied based on the given `JWT`:
```

- No `JWT`: Isolation is `Tenant` and value for the tenant is the provider account.
- `JWT` without `user_id`: Isolation is `Tenant` and value for the tenant is the `zid` of the `JWT`
- `JWT` with `user_id`: Isolation is `TenantUser` and values are taken from `zid` and `user_id`.

> Also here you can overwrite the choice of the SAP Cloud SDK.

Reviewer: ```suggestion
You can override the autodetected isolation strategy.
```

> This could be necessary if you want to register a user-independent destination per tenant but your `JWT` contains a `user_id`.

Reviewer: ```suggestion
This could be necessary if you want to register a destination that cares only about `Tenant` while your `JWT` contains both a `tenant` and a `user_id` which would make it `TenantUser` by default.
```

---
title: Destination Cache and Isolation
hide_title: false
hide_table_of_contents: false
sidebar_label: Destination Cache
description: This article describes how destinations are cached and how the cache entries are isolated with respect to Tenant and User.
keywords:
  - sap
  - cloud
  - sdk
  - destination
  - connectivity
  - JavaScript
  - TypeScript
---

The SAP Cloud SDK provides the option to cache destinations for efficiency reasons.
If a destination is stored in the cache the question about isolation arises.
Should the destination be cached for all users?
This guide explains the different options you have regarding the destination cache of the SAP Cloud SDK.
The general destination lookup is described in [this guide](./destination.mdx).

## Destination-Service Cache

When you request a destination from the destination-service you can improve performance by enabling the cache.
This reduces the number of calls to the destination service.
All the discussed options apply to the `execute()` and `getDestination()` method of the SAP Cloud SDK.

> The destination caching is disabled by default and you switch is on by the `useCache` flag:

Reviewer: ```suggestion
The destination caching is disabled by default and you switch it on by the `useCache` flag:
```

Reviewer: I think it's a typo.
